src/functional/iterator/embedded.py

```diff
@@ -183,39 +247,47 @@ def less(first, second):
     return first < second
 
 
-def named_range_(axis: CartesianAxis, range_: range):
+def named_range_(axis: str, range_: Iterable[int]) -> Iterable[tuple[str, int]]:
     return ((axis, i) for i in range_)
 
 
-def domain_iterator(domain: dict[CartesianAxis, range]):
-    return (
-        dict(elem)
-        for elem in itertools.product(*(named_range_(tup[0], tup[1]) for tup in domain.items()))
+def domain_iterator(domain: dict[str, range]) -> Sequence[Position]:
+    return cast(
+        Sequence[Position],
+        (
+            dict(elem)
+            for elem in itertools.product(*(named_range_(tup[0], tup[1]) for tup in domain.items()))
+        ),
     )
 
 
-def execute_shift(pos, tag, index, *, offset_provider):
+def execute_shift(
+    pos: MaybePosition, tag: str, index: int, *, offset_provider: OffsetProvider
+) -> MaybePosition:
+    assert pos is not None
     if tag in pos and pos[tag] is None:  # sparse field with offset as neighbor dimension
         new_pos = pos.copy()
         new_pos[tag] = index
         return new_pos
-    assert tag.value in offset_provider
-    offset_implementation = offset_provider[tag.value]
-    if isinstance(offset_implementation, CartesianAxis):
-        assert offset_implementation in pos
+    assert tag in offset_provider
+    offset_implementation = offset_provider[tag]
+    if isinstance(offset_implementation, Dimension):
+        assert offset_implementation.value in pos
         new_pos = pos.copy()
-        new_pos[offset_implementation] += index
+        _extracted_val_for_mypy_check = new_pos[offset_implementation.value]
+        assert isinstance(_extracted_val_for_mypy_check, int)
+        new_pos[offset_implementation.value] = _extracted_val_for_mypy_check + index
         return new_pos
     elif isinstance(offset_implementation, NeighborTableOffsetProvider):
-        assert offset_implementation.origin_axis in pos
+        assert offset_implementation.origin_axis.value in pos
         new_pos = pos.copy()
-        del new_pos[offset_implementation.origin_axis]
-        if offset_implementation.tbl[pos[offset_implementation.origin_axis], index] is None:
+        del new_pos[offset_implementation.origin_axis.value]
```

Comment:
For readability I prefer to use `dict.pop()`, but it's mostly a question of taste, so feel free to ignore.
```suggestion
        new_pos.pop(offset_implementation.origin_axis.value)
```